### PR TITLE
Cover failed assertions

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**
This adds an option `--cover-failed-assertions` that prevents the
default behaviour of coverage stopping at failed assertions by turning
assertions into skips rather than assumes (which is the default
behaviour for coverage criteria other than `assertion`, which behaves
the same with or without the flag).

This is one way to address https://github.com/diffblue/cbmc/issues/5543

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/cover.cpp
Comment (hannes-steffenhagen-diffblue) [inline src/goto-instrument/cover.cpp:307]: Btw we have to skip them rather than just leaving them be because otherwise cbmc would consider these assertions coverage goals.

File: src/cbmc/README.md
Comment (hannes-steffenhagen-diffblue) [inline src/cbmc/README.md:83]: BTW shouldn’t this file be under `doc` rather than `src`? I don’t really understand our documentation structure...

Comment (peterschrammel) [reply src/cbmc/README.md:83]: The READMEs which are in each module are pulled into here: http://cprover.diffblue.com/folder-walkthrough.html (click on `cbmc`, for example).
We decided to keep them in the source folders so that they are as close as possible to the code.

File: src/goto-instrument/cover.h
Comment (martin-cs) [inline src/goto-instrument/cover.h:27]: Is it possible to have this as an argument to `--cover`?

File: src/goto-instrument/cover.h
Comment (martin-cs) [inline src/goto-instrument/cover.h]: Thank you for this refactoring.  Finding an removing unnecessary and redundant repetition is always good.

File: src/cbmc/README.md
Comment (martin-cs) [inline src/cbmc/README.md:80]: I don't want to open a massive can of worms here but ... why is this the default?  Is this actually useful?  Given that there are options for `--no-assertions`, `--no-assumptions` and `--assert-to-assume` (whether they work or even do anything at all ... is an open question) could we make the transformation of assertions and assumptions orthogonal to `--cover`.  It feels like this should be possible and even desirable...

**Review by martin-cs — APPROVED** [review]
Approve because the refactor is good.  I would prefer if the new option could be added in to the argument for `--cover` and I would prefer it even more if we could make assertion transformation orthogonal to and preceeding coverage assertions or at least have a conversation about it but I also realise that you need to get stuff done so approve.

File: src/cbmc/README.md
Comment (hannes-steffenhagen-diffblue) [reply src/cbmc/README.md:80]: @martin-cs I think this is the default because whoever wrote this had the expectation that programs should stop at failed assertions. I suppose we could change it, but I’m always a bit wary about changing defaults.

Comment (hannes-steffenhagen-diffblue) [reply src/cbmc/README.md:80]: @martin-cs FWIW because cover treats each failed goal as something relating to coverage right now we have to do _something_ with asserts. That said, I think the alternative (just making cover cleverer about how it counts coverage) would be more desirable.

I'd consider that as something we can take a look at in terms of future work though.

Comment (martin-cs) [reply src/cbmc/README.md:80]: Yes, if you are doing a naive count then, I see the need to do something with pre-existing assertions or otherwise true assertions (generally a good thing) will count against your coverage.  Smarter counting would certainly be one way to address this.

It feels to me that making the "doing something about asserts" independent might be a desirable option as well.  I can see use-cases for converting to SKIP, to ASSUME and leaving as they are.

Then again, I am not a user of this code at the moment and do not have time to work on it so my views maybe not so important.

File: regression/validate-trace-xml-schema/check.py
Comment (peterschrammel) [inline regression/validate-trace-xml-schema/check.py]: including -> included?

File: src/goto-instrument/cover.cpp
Comment (peterschrammel) [reply src/goto-instrument/cover.cpp:307]: Why not make `--no-assertions` (aka turn-assertions-into-skips) useable with `--cover` then (as @martin-cs  suggests)? That would make the behaviour more transparent to the user.

File: src/cbmc/cbmc_parse_options.h
Comment (thomasspriggs) [inline src/cbmc/cbmc_parse_options.h]: :grey_question: Is this additional spacing accidental?

**Comment by hannes-steffenhagen-diffblue** [comment]
@peterschrammel `--no-assertions` applies to `goto_check` assertions, it only turns user provided assertions into skips. What we'd want for cover is to turn _all_ non-cover related assertions into skips, user added or not. Making it work differently depending on whether or not `--cover` is passed would be confusing.

That said, maybe the better solution would be to make sure cover only checks cover-added assertions for coverage metrics instead of this skip-workaround...

**Comment by peterschrammel** [comment]
...unless you use `--cover assertion`.

**Comment by hannes-steffenhagen-diffblue** [comment]
I don’t think the behaviour is any different for `--cover assertion`? It still only removes user assertions (from what I can tell `--cover assertions` doesn’t care specifically about this option at all, this is all happening in goto check), and the original complaint this is meant to address is that coverage changes based on whether or not you pass in the various `*-check` flags (which I think may well be expected behaviour in some circumstances, but is clearly not the only reasonable behaviour).

It’s just that (only) removing user assertions is actually what we want for `--cover assertion` I guess.

File: regression/cbmc/cover-failed-assertions/test-no-failed-assertions.desc
Comment (thomasspriggs) [inline regression/cbmc/cover-failed-assertions/test-no-failed-assertions.desc:11]: :pick: It would be nice to have a comment in each of `.desc` files explaining what the test is intended to test. Even something pretty short could make this a while lot easier to maintain.

File: src/goto-instrument/cover.h
Comment (thomasspriggs) [inline src/goto-instrument/cover.h:32]: :pick: Because of the preceding refactor, this option now applies to `jdiff` and `goto-diff` as well as `cbmc`. This means that we should probably test that it works with those entry points.

File: src/goto-instrument/cover.h
Comment (thomasspriggs) [inline src/goto-instrument/cover.h:27]: :pick: If we keep this as a separate option from `--cover` then we should check that `--cover-failed-assertions` is not passed without `--cover` and exit with a suitable error message in the case of that particular invalid usage.

**Review by thomasspriggs — APPROVED** [review]
Seems reasonable. If it was up to me, I would squash the commits fixing the tests back into the commit where they were added. This is because it could make use of `git bisect` easier, if we ever need to use it. Please fix the spelling error noted by Peter before this is merged.

**Review by peterschrammel — APPROVED** [review]
I understand that this needs a quick fix. But I don't think that the current solution will be the last. The options around goto check assertions, user assertions, cover, the various -no-X options, etc need to be made consistent so that they can be used in a clearly understandable and orthogonal way without subtle side effects.

**Comment by martin-cs** [comment]
I realise this is easy for me to say as 1. it's not me doing the work and 2. it's not me in need of the urgent fix but... +1 agree with @peterschrammel .  This has been a mess for too long.  I will leave it up to others (those that 1 and 2 apply to) to decide how this PR should relate to that work.

**Pull request merged**
